# Workshop 1
## Initial Data Profiling

Initial dataset profiling.

In [12]:
import pandas as pd

df = pd.read_csv('candidates.csv', sep=';')
df.head()

,First Name,Last Name,Email,Application Date,Country,YOE,Seniority,Technology,Code Challenge Score,Technical Interview Score
0,Bernadette,Langworth,leonard91@yahoo.com,2021-02-26,Norway,2,Intern,Data Engineer,3,3
1,Camryn,Reynolds,zelda56@hotmail.com,2021-09-09,Panama,10,Intern,Data Engineer,2,10
2,Larue,Spinka,okey_schultz41@gmail.com,2020-04-14,Belarus,4,Mid-Level,Client Success,10,9
3,Arch,Spinka,elvera_kulas@yahoo.com,2020-10-01,Eritrea,25,Trainee,QA Manual,7,1
4,Larue,Altenwerth,minnie.gislason@gmail.com,2020-05-20,Myanmar,13,Mid-Level,Social Media Community Management,9,7


## 1. Rows and columns

In [22]:
print(f"Rows: {df.shape[0]}")
print(f"Colmns: {df.shape[1]}")
print()
print("Column names:")
print(list(df.columns))

Rows: 50000
Colmns: 10

Column names:
['First Name', 'Last Name', 'Email', 'Application Date', 'Country', 'YOE', 'Seniority', 'Technology', 'Code Challenge Score', 'Technical Interview Score']


## 2. Data types

In [14]:
df.dtypes

First Name                     str
Last Name                      str
Email                          str
Application Date               str
Country                        str
YOE                          int64
Seniority                      str
Technology                     str
Code Challenge Score         int64
Technical Interview Score    int64
dtype: object

## 3. Null values

In [15]:
df.isnull().sum()

First Name                   0
Last Name                    0
Email                        0
Application Date             0
Country                      0
YOE                          0
Seniority                    0
Technology                   0
Code Challenge Score         0
Technical Interview Score    0
dtype: int64

## 4. Duplicate records

In [16]:
print("Duplicate rows:", df.duplicated().sum())
print("Duplicate emails (potential candidates who applied more than once):", df['Email'].duplicated().sum())

Duplicate rows: 0
Duplicate emails (potential candidates who applied more than once): 167


## 5. Unique values in relevant categorical attributes

In [17]:
for col in ['Country', 'Seniority', 'Technology']:
    print(f"{col}: {df[col].nunique()} unique values")

print()
print("Seniority Levels:")
print(sorted(df['Seniority'].unique()))

Country: 244 unique values
Seniority: 7 unique values
Technology: 24 unique values

Seniority Levels:
['Architect', 'Intern', 'Junior', 'Lead', 'Mid-Level', 'Senior', 'Trainee']


## 6. Application date range

In [18]:
df['Application Date'] = pd.to_datetime(df['Application Date'])
print("Min date:", df['Application Date'].min().date())
print("Max date:", df['Application Date'].max().date())

Min date: 2018-01-01
Max date: 2022-07-04


## 7. Score range and descriptive statistics

In [19]:
df[['YOE', 'Code Challenge Score', 'Technical Interview Score']].describe()

,YOE,Code Challenge Score,Technical Interview Score
count,50000.000000,50000.000000,50000.000000
mean,15.286980,4.996400,5.003880
std,8.830652,3.166896,3.165082
min,0.000000,0.000000,0.000000
25%,8.000000,2.000000,2.000000
50%,15.000000,5.000000,5.000000
75%,23.000000,8.000000,8.000000
max,30.000000,10.000000,10.000000


## 8. Business rule preview
For exploratory purposes only.

In [23]:
hired = (df['Code Challenge Score'] >= 7) & (df['Technical Interview Score'] >= 7)
print("Hired candidates:", hired.sum())
print(f"Hiring rate: {hired.mean()*100:.2f}%")

Hired candidates: 6698
Hiring rate: 13.40%


## Main findings

* The dataset has **50,000 rows and 10 columns**, with no null values in any column.
* The file uses `;` as a separator (not `,`) this needs to be kept in mind during the Extract step.
* There are no fully duplicate rows, but **167 emails are repeated** → some candidates applied more than once. This confirms that the grain of the dimensional model must be "one application", not "a unique candidate".
* There are **244 distinct countries** and **24 distinct technologies**.
* `Seniority` has **7 levels**: Architect, Intern, Junior, Lead, Mid-Level, Senior, Trainee.
* `YOE` ranges from **0 to 30 years** (average ~15.3), and is not correlated with Seniority (they are independent attributes).
* Both scores (`Code Challenge Score` and `Technical Interview Score`) range from **0 to 10**.
* Application dates range from **2018-01-01 to 2022-07-04**.
* Under the HIRED business rule, the overall hiring rate is **~13.4%**.
